tests/console/docker.pm

```diff
@@ -50,47 +50,62 @@ sub run {
     validate_script_output("docker search --no-trunc opensuse", sub { m/This project contains the stable releases of the openSUSE distribution/ });
 
     # images can be pulled from the Docker Hub
-    #   - pull minimalistic alpine image of declared version
+    #   - pull minimalistic alpine image of declared version using tag
     #   - https://store.docker.com/images/alpine
     my $alpine_image_version = '3.5';
     assert_script_run("docker pull alpine:$alpine_image_version", 300);
-    assert_script_run("docker images -q alpine:$alpine_image_version");
-    #   - pull hello-world image, typical docker demo image
+    assert_script_run("docker images alpine:$alpine_image_version | grep alpine");
+    #   - pull typical docker demo image without tag. Should be latest.
     #   - https://store.docker.com/images/hello-world
     assert_script_run("docker pull hello-world", 300);
-    assert_script_run("docker images -q hello-world:latest");
+    assert_script_run("docker images hello-world | grep latest");
 
     # local images can be listed
-    validate_script_output("docker images", sub { m/alpine/ });
+    my $cmd_docker_images = 'docker images';
+    my $listed_images     = script_output($cmd_docker_images);
+    unless ($listed_images =~ m/alpine/ && $listed_images =~ m/hello-world/) {
+        die("error: missing image: $cmd_docker_images");
+    }
+
 
     # containers can be spawned
-    validate_script_output("docker run --name test_1 hello-world",                                                     sub { m/Hello from Docker/ });
-    validate_script_output("docker run --name test_2 alpine:$alpine_image_version /bin/echo Hello world",              sub { m/Hello world/ });
-    validate_script_output("docker ps -a",                                                                             sub { m/test_1/ });
-    validate_script_output("docker ps -a",                                                                             sub { m/test_2/ });
-    validate_script_output("docker run --name test_ephemeral --rm alpine:$alpine_image_version /bin/echo Hello world", sub { m/Hello world/ });
+    assert_script_run('docker run --name test_1 hello-world | grep "Hello from Docker\!"');
+    assert_script_run(qq#docker run --name test_2 alpine:$alpine_image_version /bin/echo Hello world | grep "Hello world"#);
+    assert_script_run(qq#docker run --name test_ephemeral --rm alpine:$alpine_image_version /bin/echo Hello world | grep "Hello world"#);
```

Comment:
A more common style is to use differing start-/end markers. I recommend as also done by others elsewhere to use `qq{...}`

Reply:
Same here, for making the test test, I would use `docker container logs $container_name`. Otherwise, this isn't much of an improvement, unless there's an advantage of replacing `validate_script_output`.

Reply:
@okurz, Since we are using perl always against a terminal, the only character that we would never send to the SUT is '#', the comment character, so I think this would be better. And '{', 
'}' are characters very used on regexp, find command, etc...

But, if there is a concern, I can change it.

Reply:
@drpaneas, it is again about false negatives, docker container logs will not give you a $?=1, and validate_script_output only validated one of the tests. If you use both in the same regexpt, you would use '||', so false negative when one of the test are missing. And executing the command three times to apply a regexp for the three different container is not something that a user would do.
The most similar way a user would behave, is execute it once, verify output. And that's it the better way I have found for it.

Reply:
Sure, I guess most of the users would do it this way you describe, so go for it.

Personally I would prefer to check the logs:

```bash
panos@panos:~> docker container logs test_1  2>&1 | grep 'Hello from Docker!'
Hello from Docker!
panos@panos:~> echo $?
0

panos@panos:~> docker container logs test_1  2>&1 | grep 'Hello from Dockesdfr!'
panos@panos:~> echo $?
1
```